Gallery Drag and Drop Reordering [Using Draggable alternative]

caxco93 opened the pull request:
## Description
**[IN PROGRESS]**
This PR looks to implement drag and drop reordering for the Gallery Block items.
This alternative tries to accomplish this by using the [Draggable](https://github.com/WordPress/gutenberg/blob/53b10d3ffec11b4d6eacc57f98639eab86852ce4/packages/components/src/draggable/README.md) component.

![dnddraggable](https://user-images.githubusercontent.com/3190666/46255904-da206880-c468-11e8-9163-20a27cbeaeba.gif)

Alternative to #7946
Addresses Issue #743

## Known Issues:
The mouse down behaviour sticks after dropping an element. You have to click somewhere to have it behave like normal

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

caxco93 committed:
Added Draggable handle to Gallery Images

caxco93 committed:
Wrapped Dashicon with a `span`

nosolosw commented:
@caxco93 pinged me about this PR. Going to wrap up the day, but wanted to post here what I've found to date. This is what I've seen: 

* I try to drag a gallery image grabbing the `move` dashicon. To be specific: _mousedown_ and _mousemove_ events, not _mousedown_ and _mouseup_ events.
* Draggable's `componentWillUnmount` is called, which executes the `resetDragState`.
* Draggable's `dragStart` handler is fired.

It seems like the next step would be to understand why `draggable` is unmounted while `GalleryImage` hasn't been updated or unmounted. I'm not familiar with the Gallery codebase, but an idea worth looking into is reviewing any CSS that may change the DOM nodes of the gallery elements (its visibility, pointer-events properties, etc).

nosolosw commented:
Actually, forget about what I said yesterday - I was confused by two different `Draggable` instances. This is what is happening: 

- The `Draggable` component attached to the gallery image is correctly bound and its `dragStart` method starts correctly.
- At some point later, the block itself [prevents](https://github.com/WordPress/gutenberg/blob/62c527f89ac28467db6a8cbb50a188f4377a897a/packages/editor/src/components/block-list/block.js#L262) any further dragging events from firing.

Removing the `preventDrag` handler makes this work: 

![peek 2018-09-25 16-41](https://user-images.githubusercontent.com/583546/46022116-48b89b80-c0e2-11e8-84ff-94c71cddb616.gif)

I'm unsure about the rationale for the `preventDrag` handler and the component that bounds it (`IgnoreNestedEvents`) is quite obscure. I'll need some time to understand why is there and how can we do things differently. In the meantime, I hope this unblocks this PR and work can continue to implement the actual dropping mechanism.

nosolosw commented:
@caxco93 I've just reviewed my workload for the coming days/weeks and I've realized I need to focus on other tasks first. Would you mind taking over from what I've left and investigating how to fix the `IgnoreNestedEvents` thing?

caxco93 commented:
Yes @nosolosw absolutely I will continue this. Thank you very much for all the info!

caxco93 committed:
Implemented and added Reorder Dropzones

There are now Reorder Dropzones around the Gallery Items.
These require a `handleDrop` prop which perform the reordering.

caxco93 committed:
Removed only to try out the changes

This commit must not be present when merging!

caxco93 committed:
Changed selectedImage from index based to id based

This helps keep the selection after reordering

caxco93 committed:
Changed the id reference for the draggable clone

It now uses the Gallery Image itself instead of the
whole Gallery Item because of styling issues

caxco93 committed:
Set the correct styling for the Drag Handle Icon

caxco93 commented:
Hey @nosolosw, would you mind taking over this now that it's on scope as of #13113? I won't be able to work on this for now.

nosolosw commented:
Going to close this at the moment as there is a master issue for this https://github.com/WordPress/gutenberg/issues/743. We can always reopen later.
